### PR TITLE
New config style

### PR DESCRIPTION
**Pull request by marian-vignau**

### CODE REVIEW

File: recordium/config.py [hunk 1]
```diff
@@ -54,17 +48,24 @@ def __init__(self):
...
         logger.debug("Loaded: %s", self.data)
 
-    def get(self, key):
+    def __getattr__(self, key):
         return self.data.get(key, self._config_options[key])
 
-    def set(self, key, value):
-        self.data[key] = value
+    def __setattr__(self, key, value):
+        if key in self._config_options.keys():
```
Comment (facundobatista) [inline recordium/config.py]: To check the "in" you don't need to retrieve the dict keys

File: recordium/config.py [hunk 2]
```diff
@@ -54,17 +48,24 @@ def __init__(self):
...
+            if self.__getattr__(key) != value:
```
Comment (facundobatista) [inline recordium/config.py]: Why not getting the value directly here from self.data?

Let's do that, accessing with .get(), no matter if the external code is setting the default value, we're ok with that.

File: tests/test_config.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
+# Copyright 2016-2017 Facundo Batista, María Andrea Vignau
```
Comment (facundobatista) [inline tests/test_config.py]: Copyright should be just 2017, as this file didn't exist in 2016.

Also, copyright owner is me, you are the author of the code, not the copyright holder.

File: tests/test_config.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    """Tests for config module."""
+
+    def setUp(self):
+        """Set up. Creates a fake config"""
+        self.config_test = _Config()
+
+    def test_set(self):
+        """Basic test: setting and getting a key"""
+        newvalue = 15
+        self.assertNotEqual(self.config_test.POLLING_TIME, newvalue)
```
Comment (facundobatista) [inline tests/test_config.py]: when you are just checking that a precondition for the test is True, don't use self.assert*, just assert, like:

    assert self.config_test.POLLING_TIME != newvalue

File: tests/test_config.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        self.config_test.POLLING_TIME = newvalue
+        self.assertEqual(self.config_test.POLLING_TIME, newvalue)
+
+    def test_attribute_error(self):
+        """Try to access an inexistent attribute to raise an error"""
+        self.assertRaises(AttributeError, exec("self.config_test.SARASA = 1"))
```
Comment (facundobatista) [inline tests/test_config.py]: I really don't like this "exec" here, can't you use setattr() or something?

File: tests/test_config.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        cfg = _Config()
+        self.assertNotEqual(cfg.USER_ALLOWED, newvalue)
+
+        self.config_test.USER_ALLOWED = newvalue
+        self.config_test.save()
+
+        cfg = _Config()
+        self.assertEqual(cfg.USER_ALLOWED, newvalue)
+
+    def tearDown(self):
```
Comment (facundobatista) [inline tests/test_config.py]: tearDown method should be just after the setUp one, above

File: tests/test_config.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+        """Remove fake config file"""
+        if os.path.exists(config.FILEPATH):
+            os.remove(config.FILEPATH)
+
+
+# to allow the creation of a fake config file
+parent_dir = os.path.dirname(os.path.realpath(__file__))
+config.FILEPATH = os.path.join(parent_dir, "test.cfg")
```
Comment (facundobatista) [inline tests/test_config.py]: don't put the temp test file in a standard project directory; use a "tmp" one inside the project, or just "/tmp"

File: tests/test_config.py [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+_Config = config.__dict__['_Config']
```
Comment (facundobatista) [inline tests/test_config.py]: why this is needed?

File: tests/test_config.py [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+        cfg = _Config()
+        self.assertEqual(cfg.USER_ALLOWED, newvalue)
+
+    def tearDown(self):
+        """Remove fake config file"""
+        if os.path.exists(config.FILEPATH):
+            os.remove(config.FILEPATH)
+
+
+# to allow the creation of a fake config file
```
Comment (facundobatista) [inline tests/test_config.py]: if you do these kind of actions at module level, they will happen once for all the tests. DO NOT. Do everything you need for the test in the setUp method, and clean it (if needed) in the tearDown.

Optimization is not important here. What is really important is to be sure that every test start FRESH

**Review by facundobatista — CHANGES_REQUESTED** [review]
I like the general idea of the branch, thanks!

But please fix the marked things.

Also, you should provide a simple way to run the test(s), like a "test" file

File: tests/test_config.py (hunk 8 above)
Comment (marian-vignau) [reply tests/test_config.py]: _Config is private. I can't access it like any normal class.
This is a work around.

File: tests/test_config.py (hunk 5 above)
Comment (marian-vignau) [reply tests/test_config.py]: Why do you dislike it?
The disadvantage of using  exec is on security side. This is not an issue in a test

File: recordium/config.py [hunk 10]
```diff
@@ -52,8 +52,8 @@ def __getattr__(self, key):
         return self.data.get(key, self._config_options[key])
 
     def __setattr__(self, key, value):
-        if key in self._config_options.keys():
-            if self.__getattr__(key) != value:
+        if key in self._config_options:
+            if self.data.get(key, self._config_options[key]) != value:
```
Comment (facundobatista) [inline recordium/config.py]: What is the concept behind "not storing if the value is different of what the config has but the same than the default"?

Why not just "store if the value is different of what is already stored in the config"?

File: tests/test_config.py (hunk 5 above)
Comment (facundobatista) [reply tests/test_config.py]: Because there is no need of calling it. "exec" is a corner case instruction for when you need to execute some code in particular conditions. Why not do something more simple?

File: tests/test_config.py [hunk 11]
```diff
@@ -25,38 +25,32 @@ class TestConfig(TestCase):
     def setUp(self):
         """Set up. Creates a fake config"""
         self.config_test = _Config()
+        if os.path.exists(config.FILEPATH):
```
Comment (facundobatista) [inline tests/test_config.py]: Cleanup should not be done in the setUp, but in the tearDown. You need to END the tests with everything clean.

File: tests/test_config.py [hunk 12]
```diff
@@ -25,38 +25,32 @@ class TestConfig(TestCase):
...
+            os.remove(config.FILEPATH)
```
Comment (facundobatista) [inline tests/test_config.py]: Don't do this kind of action on "config.FILEPATH", it looks like you're deleting the real file.

Bring the change you're doing globally below (as I said in the last review), do that here, and operate on the temp file.

Example (untested):
```
    temp_file = "/tmp/blah"
    config.FILEPATH = temp_file  # but you need to restore it to the real value, as part of the cleanup, see mock.patch()
    
    os.remove(temp_file)  # later in tearDown
```

File: tests/test_config.py (hunk 9 above)
Comment (facundobatista) [reply tests/test_config.py]: You're still doing stuff here globally!

File: tests/test_config.py (hunk 8 above)
Comment (facundobatista) [reply tests/test_config.py]: _Config is private, but you can access it anyway. This is Python, not Java. Do `config._Config`.

**Review by facundobatista — CHANGES_REQUESTED** [review]

File: recordium/config.py [hunk 13]
```diff
@@ -53,11 +53,14 @@ def __getattr__(self, key):
 
     def __setattr__(self, key, value):
         if key in self._config_options:
-            if self.data.get(key, self._config_options[key]) != value:
+            if self.data.get(key) != value:
```
Comment (facundobatista) [inline recordium/config.py]: You'll need to change something here, I just noticed.

You can't use the .get(), otherwise you're forbidding me to store a config var in None.

Example, let's say that the config has a key called "foo", with default in "bar". So far, it's empty. I want to store None in "foo", the code will do `self.data.get(key) != value`, being `self.data.get(key)` equal None (because the key is not there), and `value` is None, so it won't store my value (and in the future it will use the default, "bar").

So, in this case you actually need to check "if key in self.data"... if yes, check the value, if not, store it.

File: recordium/config.py [hunk 14]
```diff
@@ -53,11 +53,14 @@ def __getattr__(self, key):
...
                 self.data[key] = value
                 self._need_save += 1
         else:
-            super(_Config, self).__setattr__(key, value)
+            if key.upper() != key:
```
Comment (facundobatista) [inline recordium/config.py]: What is this .upper() here? Why are you checking if the attribute is uppercase? using lowercase should behave the same than using a non-existant uppercase key, and it should raise AttributeError.

File: tests/test_config.py [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+        """Set up. Patch FILEPATH. Creates a fake config"""
+        self.config_patch = mock.patch("recordium.config.FILEPATH")
+        config.FILEPATH = self.config_patch.start()
+        config.FILEPATH = "/tmp/recordium.cfg"
+        self.config_test = config._Config()
+
+    def tearDown(self):
+        """Remove temporary config file. """
+        if os.path.exists(config.FILEPATH):
+            os.remove(config.FILEPATH)
```
Comment (facundobatista) [inline tests/test_config.py]: You're still calling os.remove on config.FILEPATH!!!

In the last review I told you to not do it :(

File: tests/test_config.py [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+from recordium import config
+
+class TestConfig(TestCase):
+    """Tests for config module."""
+
+    def setUp(self):
+        """Set up. Patch FILEPATH. Creates a fake config"""
+        self.config_patch = mock.patch("recordium.config.FILEPATH")
+        config.FILEPATH = self.config_patch.start()
+        config.FILEPATH = "/tmp/recordium.cfg"
```
Comment (facundobatista) [inline tests/test_config.py]: I don't think mock.patch is used correctly in this case, but I'd need to do some tests, which will come after the rest is fixed/solved.

**Review by facundobatista — CHANGES_REQUESTED** [review]

File: recordium/config.py (hunk 14 above)
Comment (marian-vignau) [reply recordium/config.py]: `__setattr__` is called every time a new attribute is added to the instance
So, if I dont't make this exception, and check this, I couldn't use self.data = pickle.load(fh), to create a new attribute in the function code.

File: tests/test_config.py (hunk 15 above)
Comment (marian-vignau) [reply tests/test_config.py]: It is because config.FILEPATH is patched, it ceased only in the next line, on `self.config_patch.stop()`

File: recordium/config.py [hunk 17]
```diff
@@ -43,29 +35,40 @@ class _Config(object):
         'USER_ALLOWED': None,
         'COL_ORDER': []
     }
+    _need_save = 0
```
Comment (facundobatista) [inline recordium/config.py]: This attribute should not be a class one, but a instance one (set it inside __init__). Also, name it "_needs_save" or "_save_needed"

File: recordium/config.py [hunk 18]
```diff
@@ -43,29 +35,40 @@ class _Config(object):
...
-        self.data[key] = value
+            self._data = pickle.load(fh)
+        logger.debug("Loaded: %s", self._data)
+
+    def __getattr__(self, key):
+        return self._data.get(key, self._config_options[key])
+
+    def __setattr__(self, key, value):
+        if key in self._config_options:
+            if not key in self._data or self._data[key] != value:
```
Comment (facundobatista) [inline recordium/config.py]: It's simpler to read "if key not in self...."

File: recordium/config.py [hunk 19]
```diff
@@ -43,29 +35,40 @@ class _Config(object):
...
+                self._data[key] = value
+                self._need_save += 1
+        else:
+            if key.startswith('_'):
+                super(_Config, self).__setattr__(key, value)
```
Comment (facundobatista) [inline recordium/config.py]: In Py3 you can omit the super parameters: super().__setattr__...

File: tests/test_config.py (hunk 15 above)
Comment (facundobatista) [reply tests/test_config.py]: Please, store the tmp filename in an instance attribute, self.temp_filename, and call os.remove() on that. 

Otherwise, it's error prone: say you remove by accident a patcher line, then you run the tests, and it removed your REAL config!!

**Review by facundobatista — CHANGES_REQUESTED** [review]
Final round of changes?

**Comment by facundobatista** [comment]
Thanks! Excellent work!
